ugs-platform/ugs-platform-ugscore/src/main/java/com/willwinder/ugs/nbp/core/actions/OutlineAction.java

```diff
@@ -0,0 +1,143 @@
+        this.backend = CentralLookup.getDefault().lookup(BackendAPI.class);
+        this.backend.addUGSEventListener(this);
+
+        putValue("iconBase", ICON_BASE);
+        putValue(SMALL_ICON, ImageUtilities.loadImageIcon(ICON_BASE, false));
+        putValue("menuText", LocalizingService.OutlineTitle);
+        putValue(NAME, LocalizingService.OutlineTitle);
+        setEnabled(isEnabled());
+    }
+
+    @Override
+    public void UGSEvent(UGSEvent cse) {
+        java.awt.EventQueue.invokeLater(() -> setEnabled(isEnabled()));
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return backend != null && backend.isIdle() && backend.getGcodeFile() != null;
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+        try {
+            File gcodeFile = backend.getGcodeFile();
+            List<GcodeCommand> gcodeCommands = generateOutlineCommands(gcodeFile);
+            backend.getController().queueStream(new SimpleGcodeStreamReader(gcodeCommands));
+            backend.getController().beginStreaming();
+        } catch (Exception ex) {
+            GUIHelpers.displayErrorDialog(ex.getLocalizedMessage());
+        }
+    }
+
+    public List<GcodeCommand> generateOutlineCommands(File gcodeFile) throws IOException, GcodeParserException {
+        List<LineSegment> gcodeLineList = parseGcodeLinesFromFile(gcodeFile);
+
+        // We only care about carving motion, filter those commands out
+        List<PartialPosition> pointList = gcodeLineList.parallelStream()
+                .filter(lineSegment -> !lineSegment.isFastTraverse())
+                .flatMap(lineSegment -> {
+                    // We map both the start and end points in MM
+                    PartialPosition start = PartialPosition.fromXY(lineSegment.getStart().getPositionIn(UnitUtils.Units.MM));
+                    PartialPosition end = PartialPosition.fromXY(lineSegment.getEnd().getPositionIn(UnitUtils.Units.MM));
+                    return Stream.of(start, end);
+                })
+                .distinct()
+                .collect(Collectors.toList());
+
+        List<PartialPosition> outline = MathUtils.generateConvexHull(pointList);
+        return outline.stream()
+                .map(point -> new GcodeCommand(GcodeUtils.generateMoveToCommand(Code.G0.name(), point)))
+                .collect(Collectors.toList());
+    }
+
+    private List<LineSegment> parseGcodeLinesFromFile(File gcodeFile) throws IOException, GcodeParserException {
+        List<LineSegment> result;
+
+        GcodeViewParse gcvp = new GcodeViewParse();
+        try (IGcodeStreamReader gsr = new GcodeStreamReader(gcodeFile)) {
+            result = gcvp.toObjFromReader(gsr, ARC_SEGMENT_LENGTH);
+        } catch (GcodeStreamReader.NotGcodeStreamFile e) {
+            List<String> linesInFile = VisualizerUtils.readFiletoArrayList(gcodeFile.getAbsolutePath());
+            result = gcvp.toObjRedux(linesInFile, ARC_SEGMENT_LENGTH);
```

Comment:
Maybe the back end shouldn't generate a file event until the initial parsing to avoid this edge case.

Reply:
This code is duplicated in several places and should probably be extracted in some way. Is it ok if leave this as it is until we have the time to fix this?

Reply:
Yeah, just thinking out loud